text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+refer to a stable location, i.e. `/platform/deps/assets/…`. These dependencies
+could then be accessed by the buildpack directly, and the buildpack doesn’t
+need to care how they were added to the container.
+
+This document does not specify how the dependency asset folder should be
+provided to buildpacks but here are some possibilities:
+
+1. External. The dependency assets can be managed outside of the buildpack
+   lifecycle. This allows for users to manually pull in the dependency assets
+   they would like when they would like them. It also outlives the buildpack
+   lifecycle so assets could be shared across builds for greater caching
+   benefit. The assets can then be mapped into the build container at the
+   specified location using a volume mount to `pack build`.
+1. Via builder. The builder could come with assets included under
+   `/platform/deps/assets`. This would likely require a builder with many
+   buildpacks to update very frequently though, so may not be the best idea for
+   the average case, however, it could be very useful as a way to distribute
+   dependencies in offline environments.
+1. Via the platform. Platforms may choose to offer enhanced functionality to
+   more easily distribute dependencies. In the end, the platform just needs to
+   ensure that the dependency metadata is available to the buildpack in the
+   required location.
+
+## Implementation
+### Tools
+To start the implementation, we’ll create some tooling to work with the new
+metadata format. We’ll create some basic tools for authors to add metadata,
+update metadata, and package metadata into compressed archives and images. For
+buildpack users, we’ll create some tooling to fetch metadata, overlay metadata
+from multiple sources, pre-cache assets from metadata, and feed metadata and
+pre-cached assets into `pack build`.
+
+### Library Enhancement
+The next step will be to add new functionality in `libpak` and `packit` for
+consuming metadata from the defined metadata location and for consuming
+dependency assets from this metadata. This can be done in an additive way so
+that existing functionality continues to be supported. Support for
+`buildpack.toml` dependency metadata can be deprecated and removed on a
+separate timeline.
+
+### Update Buildpacks
+When we have library support, we’ll be able to update the Paketo buildpacks
+that are installing dependencies to use the metadata provided. For example, we
+can update the Bellsoft Liberica or Amazon Corretteo buildpacks to install
+their JRE/JDK dependencies using this new metadata.
+
+We propose a flag of `BP_EXTERNAL_METADATA_ENABLED` which defaults to `false`
+for use as buildpacks are being converted. In the default state, this flag
+tells a buildpack to use the metadata included with buildpack.toml. When set to
+`true`, a buildpack should use the new metadata. This can provide a way for
+users to test the new functionality without impacting existing users.
```

Comment:
Could we make this more smoothly, by piggy backing on the `BP_DEPENDENCY_METADATA` environment variable, i.e. when it is set use external metadata if not continue with `buildpack.toml` based metadata?

Reply:
This would work, except I think it makes for a poor UX because you have to opt-in to the new system by knowing where the files would be on the filesystem.

I think it's a better UX to ask the user to provide `BP_EXTERNAL_METADATA_ENABLED` than `BP_DEPENDENCY_METADATA=/platform/deps/metadata`

Reply:
I guess my comment was made in the context of not defaulting the path, i.e. if used in a context with external metadata, the variable would be set without the user needing to do anything.

Reply:
Yeah, I think this conversation is fairly-coupled to the other conversation above about who sets the environment variables - whether they are required during `Detect` or not.